// micro-eth-signer/esm/kzg.js
import { bitLen, bytesToNumberBE, numberToBytesBE } from '@noble/curves/abstract/utils';
import { bls12_381 as bls } from '@noble/curves/bls12-381';
import { sha256 } from '@noble/hashes/sha256';
import { bytesToHex, utf8ToBytes } from '@noble/hashes/utils';
import { add0x, hexToNumber, strip0x } from "./utils.js";
/*
KZG for [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844).

Docs:
- https://github.com/ethereum/c-kzg-4844
- https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/polynomial-commitments.md

TODO(high-level):
- data converted into blob by prepending 0x00 prefix on each chunk and ends with 0x80 terminator
  - Unsure how generic is this
  - There are up to 6 blob per tx
  - Terminator only added to the last blob
- sidecar: {blob, commitment, proof}
- Calculate versionedHash from commitment, which is included inside of tx
- if 'sidecars' inside of tx enabled:
  - envelope turns into 'wrapper'
  - rlp([tx, blobs, commitments, proofs])
  - this means there are two eip4844 txs: with sidecars and without

TODO(EIP7594):
https://eips.ethereum.org/EIPS/eip-7594
compute_cells_and_kzg_proofs(cells, proofs, blob);
recover_cells_and_kzg_proofs(recovered_cells, recovered_proofs, cell_indices, cells, num_cells);
verify_cell_kzg_proof_batch(commitments_bytes, cell_indices, cells, proofs_bytes, num_cells);
*/
const { Fr, Fp12 } = bls.fields;
const G1 = bls.G1.ProjectivePoint;
const G2 = bls.G2.ProjectivePoint;
const BLOB_REGEX = /.{1,64}/g; // TODO: is this valid?
function parseScalar(s) {
    if (typeof s === 'string') {
        s = strip0x(s);
        if (s.length !== 2 * Fr.BYTES)
            throw new Error('parseScalar: wrong format');
        s = BigInt(`0x${s}`);
    }
    if (!Fr.isValid(s))
        throw new Error('parseScalar: invalid field element');
    return s;
}
function formatScalar(n) {
    return add0x(bytesToHex(numberToBytesBE(n, Fr.BYTES)));
}
function isPowerOfTwo(x) {
    return (x & (x - 1)) === 0 && x !== 0;
}
function reverseBits(n, bits) {
    let reversed = 0;
    for (let i = 0; i < bits; i++, n >>>= 1)
        reversed = (reversed << 1) | (n & 1);
    return reversed;
}
// FFTish stuff, reverses bit in index
function bitReversalPermutation(values) {
    const n = values.length;
    if (n < 2 || !isPowerOfTwo(n))
        throw new Error(`n must be a power of 2 and greater than 1. Got ${n}`);
    const bits = bitLen(BigInt(n)) - 1;
    const res = new Array(n);
    for (let i = 0; i < n; i++)
        res[reverseBits(i, bits)] = values[i];
    return res;
}
function computeRootsOfUnity(count) {
    if (count < 2)
        throw new Error('expected at least two roots');
    const PRIMITIVE_ROOT_OF_UNITY = 7;
    const order = BigInt(Math.log2(count));
    const power = (Fr.ORDER - BigInt(1)) / BigInt(2) ** order;
    const ROOT = Fr.pow(BigInt(PRIMITIVE_ROOT_OF_UNITY), power);
    const roots = [Fr.ONE, ROOT];
    for (let i = 2; i <= count; i++) {
        roots[i] = Fr.mul(roots[i - 1], ROOT);
        if (Fr.eql(roots[i], Fr.ONE))
            break;
    }
    if (!Fr.eql(roots[roots.length - 1], Fr.ONE))
        throw new Error('last root should be 1');
    roots.pop();
    if (roots.length !== count)
        throw new Error('invalid amount of roots');
    return bitReversalPermutation(roots);
}
function pairingVerify(a1, a2, b1, b2) {
    // Filter-out points at infinity, because pairingBatch will throw an error
    const pairs = [
        { g1: a1.negate(), g2: a2 },
        { g1: b1, g2: b2 },
    ].filter(({ g1, g2 }) => !G1.ZERO.equals(g1) && !G2.ZERO.equals(g2));
    const f = bls.pairingBatch(pairs, true);
    return Fp12.eql(f, Fp12.ONE);
}
/**
 * KZG from [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844).
 * @example
 * const kzg = new KZG(trustedSetupData);
 */
export class KZG {
    constructor(setup) {
        // Should they be configurable?
        this.FIAT_SHAMIR_PROTOCOL_DOMAIN = utf8ToBytes('FSBLOBVERIFY_V1_');
        this.RANDOM_CHALLENGE_KZG_BATCH_DOMAIN = utf8ToBytes('RCKZGBATCH___V1_');
        if (setup == null || typeof setup !== 'object')
            throw new Error('expected valid setup data');
        if (!Array.isArray(setup.g1_lagrange) || !Array.isArray(setup.g2_monomial))
            throw new Error('expected valid setup data');
        // The slowest part
        let fastSetup = false;
        if ('encoding' in setup) {
            fastSetup = setup.encoding === 'fast_v1';
            if (!fastSetup)
                throw new Error('unknown encoding ' + setup.encoding);
        }
        const G1L = setup.g1_lagrange.map(fastSetup ? this.parseG1Unchecked : this.parseG1);
        this.POLY_NUM = G1L.length;
        this.G2M = setup.g2_monomial.map(fastSetup ? this.parseG2Unchecked : this.parseG2);
        this.G1LB = bitReversalPermutation(G1L);
        this.ROOTS_OF_UNITY = computeRootsOfUnity(this.POLY_NUM);
        this.POLY_NUM_BYTES = numberToBytesBE(this.POLY_NUM, 8);
    }
    // Internal
    parseG1(p) {
        if (typeof p === 'string')
            p = G1.fromHex(strip0x(p));
        return p;
    }
    parseG1Unchecked(p) {
        if (typeof p !== 'string')
            throw new Error('string expected');
        const [x, y] = p.split(' ').map(hexToNumber);
        return G1.fromAffine({ x, y });
    }
    parseG2(p) {
        return G2.fromHex(strip0x(p));
    }
    parseG2Unchecked(p) {
        const xy
